### PR TITLE
Make document thread-safe

### PR DESCRIPTION
**Pull request by pasin**
* Assuming that CBLDatabase and CBLDictionary are thread safe.
* Apply @synchronized to make CBLDocument / CBLReadOnlyDocument thread safe.

### CODE REVIEW

File: Objective-C/CBLDocument.mm [hunk 1]
```diff
@@ -198,80 +201,102 @@ - (NSUInteger)countByEnumeratingWithState: (NSFastEnumerationState *)state
...
-            [self setC4Doc: nil];
-            return YES;
+    @synchronized (self.lock) {
+        assert(_database && _c4db);
+        
+        if (!self.exists) {
+            return createError(kCBLStatusNotFound, outError);
+        }
+        
+        C4Transaction transaction(_c4db);
```
Comment (hideki) [inline Objective-C/CBLDocument.mm]: Does not require to lock `c4db`?

File: Objective-C/CBLDocument.mm [hunk 2]
```diff
@@ -198,80 +201,102 @@ - (NSUInteger)countByEnumeratingWithState: (NSFastEnumerationState *)state
...
+                                        error: outError];
+    }
 }
 
 
 - (BOOL) deleteDocument: (NSError**)outError {
-    return [self saveWithConflictResolver: self.effectiveConflictResolver
-                                 deletion: YES
-                                    error: outError];
+    @synchronized (self.lock) {
```
Comment (hideki) [inline Objective-C/CBLDocument.mm]: Lock with `saveWithConflictResolver ` method is enough?? the same question applies to all method call `saveWithConflictResolver `.

File: Objective-C/CBLDocument.mm (hunk 1 above)
Comment (pasin) [inline Objective-C/CBLDocument.mm]: c4db is locked from the CBLDatabase when calling CBLDatabase.purge() method.

File: Objective-C/CBLDocument.mm (hunk 2 above)
Comment (pasin) [inline Objective-C/CBLDocument.mm]: There are only two and locking there seems to be easier to understand.

**Comment by pasin** [comment]
Will submit a new PR.

**Pull request closed without merging**